enclave-runtime/src/tls_ra/tls_ra_client.rs

```diff
@@ -144,7 +151,21 @@ where
 	/// Read all bytes into a buffer of given length.
 	fn read_until(&mut self, length: usize) -> EnclaveResult<Vec<u8>> {
 		let mut bytes = vec![0u8; length];
-		self.tls_stream.read(&mut bytes)?;
+		let mut read_index = 0usize;
+		debug!("Attempting to read {} bytes from stream", length);
+
+		// tls_stream.read() does not guarantee to read as many bytes as the buffer is large!
+		// Always check the returned amount of actual bytes read.
+		// Especially for large quantity of bytes, multiple calls to read() have to be done, to fill the buffer!
+		while read_index < length {
+			let read_chunk_size = self.tls_stream.read(&mut bytes[read_index..])?;
+			debug!("Read a chunk of {} bytes from stream", read_chunk_size);
+			if read_chunk_size == 0usize {
+				return Err(EnclaveError::Other("Unexpected end of stream".into()))
+			}
+			read_index += read_chunk_size;
```

Comment:
At first glance these two functions are really similar, perhaps we could use this here instead of the loop?
https://docs.rs/rustls/latest/rustls/struct.Stream.html#method.read_exact

Reply:
You are absolutely right - I didn't look at the API of the TLS stream, `read_exact` is exactly (no pun intended) the function we need, thanks for noticing that 😄 Glad I can remove all of that code then